VnodeCacheDesign.md

```diff
@@ -0,0 +1,101 @@
+
+#### Other Data Structures to Consider
+
+A radix/prefix tree is another option for caching the data.  Whether we go with this option or a hashing-based approach will be based on the results of performance testing.
+
+### Using the Cache
+
+When PrjFSKext needs to determine the virtualization root of a file (or if a file is inside of a virtualization root at all), it will look up the vnode in its cache.  
+
+At a high level, there are two possible outcomes when looking up a vnode in the cache:
+
+- The vnode exists in the cache
+- The vnode does not exist in the cache
+
+However, because vnodes are recycled, the cache will need to consider three possible states for vnodes:
+
+- The vnode exists in the cache, its `vid` has not changed, and its `vrgid` matches the `vrgid` of its virtualization root:  PrjFSKext will use the `VirtualizationRootHandle` in the cache.
+
+- The vnode exists in the cache and either its `vid` has changed or its `vrgid` no longer matches the `vrgid` of its virtualization root:  The vnode has been recycled or the `vrgid` of its virtualization root has changed. PrjFSKext needs to look up the vnode's current `VirtualizationRootHandle` and update the cache entry.
+
+- The vnode does not exist in the cache:  PrjFSKext will look up its `VirtualizationRootHandle` and insert the vnode into the cache.
+
+When PrjFSKext hits one of the cache miss scenarios (2nd and 3rd bullet points above), it can find the vnode's virtualization root by walking up the vnode tree until:
+PrjFSKext can find the new root by walking up the tree until:
+- There's a cache hit
+- It encounters a virtualization root's vnode
+- It encounters the vnode for the root of the file system
+
+### Special Case: Renames
+
+*Directory renames* 
+
+Directory renames are a special case because all of the directory's children may need to have their `VirtualizationRootHandle` updated. Rather than recording\tracking all of the parent\child relationships among the vnodes PrjFSKext will (potentially) invalidate a portion of the vnode cache:
+
+- If the rename is within the same virtualization root: All entries in the cache are still valid.
+
+- If the rename source and target are outside of any virutalization roots: All entries in the cache are still valid.
+
+- If the rename moves a directory out of a virtualization root: Increase that root's `vrgid` (thereby invalidating all entries in the cache for that root)
+
+- If the rename moves a directory from outside of any virtualization root to inside a virtualization root: Increase the global "no root" `vrgid` (thereby invalidating all entries in the cache that are not inside of a virtualization root)
+
+- If PrjFSKext is unable to determine if the rename was within the same root: Increase the `vrgid` for that vnode's virtualization root (or the "no root" `vrgid` if it was not inside a root)
+
+*File renames*
+
+When a file is renamed PrjFSKext will update its `VirtualizationRootHandle` and\or `vid` as needed.
```

Comment:
I can't think of any connection between renames and `vid` change - what did you have in mind here?

Reply:
@pmj I wasn't sure if renames would change a `vnode`'s `vid.  Based on your question though it makes me think that's not the case, and I'll remove the mention of `vid`.